### PR TITLE
[1LP][RFR] Adding ability to take screenshot on exception in WebMKS console tests and some other changes

### PR DESCRIPTION
**Pull request by kedark3**
Purpose or Intent
=================

__Extending__ WebMKS Test to take screenshot of console if any exception occurs and additional changes to the test. 

{{ pytest: cfme/tests/infrastructure/test_webmks_console.py --use-provider vsphere6-nested --use-provider vsphere65-nested}}

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit 1c14fd50ddae6c77a2612c47553f2abbcab58c2a:

`cfme/tests/infrastructure/test_webmks_console.py`:
- :red_circle: Line 5:1: **F401** *'re' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 1]
```diff
@@ -79,74 +78,51 @@ def test_webmks_vm_console(request, appliance, provider, vm_obj,
 
     request.addfinalizer(vm_console.close_console_window)
     request.addfinalizer(appliance.server.logout)
+    try:
```
Comment (quarckster) [inline cfme/tests/infrastructure/test_webmks_console.py]: **REQUIRED**
Too big "try except" block, use it where it's really required.

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 2]
```diff
@@ -79,74 +78,51 @@ def test_webmks_vm_console(request, appliance, provider, vm_obj,
...
+        # Get the login screen image, and make sure it is a jpeg file:
+        screen = vm_console.get_screen(180)
+        assert imghdr.what('', screen) == 'jpeg'
+
+        with ssh.SSHClient(hostname=vm_obj.ip_address, username=console_vm_username,
+                password=console_vm_password) as vm_ssh_client:
```
Comment (quarckster) [inline cfme/tests/infrastructure/test_webmks_console.py]: **REQUIRED**
This should be a fixture.

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 3]
```diff
@@ -79,74 +78,51 @@ def test_webmks_vm_console(request, appliance, provider, vm_obj,
...
+            vm_console.send_keys("touch blather\n")
+            wait_for(func=lambda: vm_ssh_client.run_command("ls blather", ensure_user=True) == 0,
+                delay=1, num_sec=10)
+            # if file was created in previous steps it will be removed here
+            # we will get instance of SSHResult
+            # Sometimes Openstack drops characters from word 'blather' hence try to remove
+            # file using partial file name. Known issue, being worked on.
+            command_result = vm_ssh_client.run_command("rm blather", ensure_user=True)
+            assert command_result
+    except:
```
Comment (quarckster) [inline cfme/tests/infrastructure/test_webmks_console.py]: **REQUIRED**
Please specify the exception.

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 4]
```diff
@@ -79,74 +78,51 @@ def test_webmks_vm_console(request, appliance, provider, vm_obj,
...
+        # Take a screenshot if an exception occurs
+        vm_console.switch_to_console()
+        take_screenshot("ConsoleScreenshot")
+        vm_console.switch_to_appliance()
+        raise
```
Comment (quarckster) [inline cfme/tests/infrastructure/test_webmks_console.py]: **REQUIRED**
Please specify the exception.

File: cfme/tests/infrastructure/test_webmks_console.py (hunk 1 above)
Comment (kedark3) [reply cfme/tests/infrastructure/test_webmks_console.py]: As discussed on IRC, kindly resolve this.

**Comment by kedark3** [comment]
PRT Failures look like BZ, have been reported : https://bugzilla.redhat.com/1496850

File: cfme/tests/infrastructure/test_webmks_console.py (hunk 1 above)
Comment (mshriver) [reply cfme/tests/infrastructure/test_webmks_console.py]: @kedark3 can you summarize the discussion and why it was determined this is acceptable here?

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 5]
```diff
@@ -98,55 +107,31 @@ def test_webmks_vm_console(request, appliance, provider, vm_obj,
...
-
-    vm_console.wait_for_text(text_to_find=provider.data.templates.get('console_template')
-        ['prompt_text'], timeout=200)
-
-    with ssh.SSHClient(hostname=vm_obj.ip_address, username=console_vm_username,
-            password=console_vm_password) as vm_ssh_client:
         # create file on system
         vm_console.send_keys("touch blather\n")
-        wait_for(func=lambda: vm_ssh_client.run_command("ls blather", ensure_user=True) == 0,
+        wait_for(func=lambda: ssh_client.run_command("ls blather", ensure_user=True) == 0,
```
Comment (mshriver) [inline cfme/tests/infrastructure/test_webmks_console.py]: **Optional, Recommended**

Use `func_args`, `func_kwargs`, and `fail_condition` arguments to wait_for instead of using lambda and trying to create a false in `func`.

File: cfme/tests/infrastructure/test_webmks_console.py (hunk 1 above)
Comment (kedark3) [reply cfme/tests/infrastructure/test_webmks_console.py]: @mshriver : try catch block is put that way because if test fails, any of the assertion goes wrong, I want to take the screenshot. I did not use finalizer because I do not need screenshot if test passes and hence except is generic and it will re-raise the original exception after the screenshot is taken. Hence @quarckster  approved it with suggestion to as exception "Exception" on line 132.

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 6]
```diff
@@ -98,55 +107,31 @@ def test_webmks_vm_console(request, appliance, provider, vm_obj,
...
             delay=1, num_sec=10)
         # if file was created in previous steps it will be removed here
         # we will get instance of SSHResult
         # Sometimes Openstack drops characters from word 'blather' hence try to remove
         # file using partial file name. Known issue, being worked on.
-        command_result = vm_ssh_client.run_command("rm blather", ensure_user=True)
+        command_result = ssh_client.run_command("rm blather", ensure_user=True)
         assert command_result
+
+    except Exception as e:
```
Comment (mshriver) [inline cfme/tests/infrastructure/test_webmks_console.py]: I'm willing to accept this in the test for now, but this should really be in the framework itself, so that any test using a console will generate console screenshots on exceptions.  Either in the exception handler plugin, or in vm_console module itself.

Comment (kedark3) [reply cfme/tests/infrastructure/test_webmks_console.py]: I agree to this.

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 7]
```diff
@@ -52,9 +51,21 @@ def configure_vmware_console_for_test(appliance, provider):
     settings_pg.update()
 
 
+@pytest.yield_fixture
+def ssh_client(vm_obj, provider):
+    """Provide vm_ssh_client for ssh operations in the test."""
+    console_vm_username = credentials[provider.data.templates.get('console_template')
```
Comment (mshriver) [inline cfme/tests/infrastructure/test_webmks_console.py]: I think this needs to take console_template as a fixture instead of looking it up through provider. End up with the same AttrDict, but this way its more explicit that the same template is needed both here and for the test using ssh_client.

**Comment by kedark3** [comment]
Failing because #5490 needs to be merged.

File: cfme/tests/infrastructure/test_webmks_console.py (hunk 6 above)
Comment (mshriver) [reply cfme/tests/infrastructure/test_webmks_console.py]: Thanks @kedark3, saw you did create a card for moving this to make it more widely available.

**Comment by mshriver** [comment]
@kedark3 #5490 was merged, please rebase this branch so PRT runs again. I'll merge based on results, or someone else will while I'm at QE CampX.

**Comment by kedark3** [comment]
@mshriver  Thanks for update, I have rebased this PR and will kick a PRT run and update PR comments with result.

**Comment by kedark3** [comment]
@mshriver PRT Failures look like BZ, have been reported : https://bugzilla.redhat.com/1496850
Also, Upstream seems to be failing on the same error, and skipping another provider on a known issue as per conversation I had with @psav
